### ISSUE TITLE
Add 32-bit x86 Support for Nanoserver

### ISSUE THREAD

**Issue by grochoge** [microsoft/Windows-Containers#118]
I'd like to be able to run 32-bit Windows applications in a container. 

Right now there's a 20x size gap between nanoserver (which can only run 64-bit applications) and servercore (which supports WOW64).

It would be useful to have a 32-bit x86 only version of nanoserver (if that's possible to run on a 64-bit kernel) or a version of nanoserver that has WOW64 support added.

**Comment by weijuans-msft** [comment]
@grochoge  we haven't seen many people asking this. To help us prioritize, can you share more details on the use cases?

**Comment by grochoge** [comment]
@weijuans-msft the use case is running some legacy 32-bit applications (in this case built with mingw32) in a container with smaller on-disk size than Server Core.
Really it would be nice to have some sort of intermediate between Nanoserver and Server Core. Maybe a way to install only needed features (ala Enable-WindowsOptionalFeature) on top of Nanoserver. Not sure if that's been suggested already?

**Comment by ghost** [comment]
This issue has been open for 30 days with no updates.
@weijuans-msft, please provide an update or close this issue.

**Comment by slonopotamus** [comment]
> Maybe a way to install only needed features (ala Enable-WindowsOptionalFeature) on top of Nanoserver.

That would be ideal, if users could tune image contents themselves from nano to full-blown desktop image by installing optional parts.

**Comment by ghost** [comment]
This issue has been open for 30 days with no updates.
@weijuans-msft, please provide an update or close this issue.

**Comment by ghost** [comment]
This issue has been open for 30 days with no updates.
@weijuans-msft, @judyliu-ms, please provide an update or close this issue.

**Comment by judyliu-ms** [comment]
Hi @grochoge, we are currently investigating this issue. For internal reference, the ADO tracking number is 37314246.

**Comment by ghost** [comment]
This issue has been open for 30 days with no updates.
@weijuans-msft, @judyliu-ms, please provide an update or close this issue.

**Comment by ghost** [comment]
This issue has been open for 30 days with no updates.
@weijuans-msft, @judyliu-ms, please provide an update or close this issue.

**Comment by AdaskoTheBeAsT** [comment]
hi - any news about x86 version of nano?

**Comment by fschmied** [comment]
Joining in - we, too, have the scenario mentioned above. We need to run legacy processes (in this case, a SIP for Authenticode Signing that Microsoft provides in an x86 version only) in containers and would like to use the much smaller nanoserver images (or something built on them that includes WOW64).

**Comment by Zantier** [comment]
I would like this too, in order to run nginx, which is only released for Windows in 32-bit.

**Comment by weijuans-msft** [comment]
Thanks. What other 32-bit apps that folks wanted to run on Nano Server container?

**Comment by AdaskoTheBeAsT** [comment]
Hi, by general in some industries like HVAC it is very common that component manufacturers for water heaters, coolers, fans etc provide their calculation libraries in form of dll's. Usually those companies do not have strong it departments (pun intended ;)) and they provide dlls targeted only for x86. In my professional life I saw plain dll's and also COM libraries written in VB, C++, Delphi even PowerBasic. It is very beneficial to have some interop layer written in .NET 6 on top of them and make windows container connect through some message bus (as some of them are able to calculate only one calculation in parallel) and multiply those containers with some load balancing. But right now only available solution is windows core container which is ~5GB in size - too large for simple worker container ;)

**Comment by weijuans-msft** [comment]
@AdaskoTheBeAsT Interesting! How and where do you run those Windows containers? What is the scale/size? Do you use any orchestrator like K8s?

**Comment by AdaskoTheBeAsT** [comment]
@weijuans-msft  many years ago I made such systems where such dll's were "hidden" by web apis hosted on IIS on same server - when I got back to such architecture after years I saw potential for improvements to modernize that architecture - nowadays natural direction is to do slim containers which can be instantiated on demand etc - I am not expert in kubernetes (but planning to extend my knowledge) - I thought that not only in HVAC but also other less rich industries it can be common problem - legacy systems etc which can be hidden by small facade located on slim windows container

**Comment by SamCotroneo-Wymac** [comment]
Being able to run run the official nginx for windows inside a nano container would be amazing.

**Comment by romandolinsky** [comment]
@SamCotroneo-Wymac There is windows 64bit instance of nginx in official nginx page. 
But Like was mention: nanoserver with x86 support will make our life easier.

**Comment by Zantier** [comment]
@romandolinsky I don't want to derail this issue too much, but what makes you say that there's a 64-bit Windows build of nginx on the official site?

If I go to the downloads page, and download "nginx/Windows-1.23.1", the only exe in the zip file is "nginx.exe", which you can see is 32-bit with a hex-editor (or otherwise).

**Comment by romandolinsky** [comment]
Sorry for my mistake but source which I use in past was not official.

**Comment by Sudoite** [comment]
I have a use case for x86 on Nanoserver: hosting a vulnerable x86 Windows binary in a Docker container with a small footprint for CTF competitions. The smaller footprint in theory makes learning Windows security accessible to a broader audience than if we were to require participants to have more storage readily available. I was able to get my binary working by using Process Monitor to figure out what DLLs were missing and then copy them into the appropriate locations in Nanoserver during the Docker build process. But that's a messy, host-specific workaround. Definitely would like to see this feature!

**Comment by RealFrogo** [comment]
In my company we have a lot of legacy code which relies on windows based 32bit software to be built and maintained, which cannot be easily changed, not least due to very strict admission requirements. As we are currently trying to upgrade our build servers, we want to switch to using containers as well, but want them to be as small as possible, as huge containers slow down the build processes. And to be honest, the gap in size between server core and nano server is ridiculous (if I may say so), if there are no intermediate stages available. A customizable version of the nanoserver (including 32bit support) would therefore be highly appreciated, just as it is possible with all those linux images out there. It would be great if you could make this possible, one way or the other!

**Comment by smkanadl** [comment]
Still interested in this!

**Comment by RealFrogo** [comment]
So am I!

**Comment by akrus** [comment]
We're also interested in this.

**Comment by akarshm** [comment]
This is in our backlog and will be considered a feature item for the next major release for Windows Server Containers. Unfortunately, I don't have a timeline to share at this time.

**Comment by Jens-G** [comment]
Slightly offtopic: It might be worth reconsidering the fact, that today we still have to use 32-bit installers to install runtimes on 64 bit platforms. Having e.g. a 64 bit installer for vcredist could save a lot of time for many people [trying to install the runtime in a nano-based container and only getting **exit code 3221225781** as a result](https://stackoverflow.com/questions/60348640/docker-returning-exit-code-3221225781-installing-vc-redist-x64-exe).

**Comment by malaterre** [comment]
> Thanks. What other 32-bit apps that folks wanted to run on Nano Server container?

I'd like to be able to run `vswhere.exe` from the vs_buildtools. See:

* https://github.com/microsoft/vswhere

**Comment by smkanadl** [comment]
> Thanks. What other 32-bit apps that folks wanted to run on Nano Server container?

I have some .NET code that specifically uses x86 native code. Being able to test and deploy this on the much smaller nano image would be very helpful!

**Comment by RealFrogo** [comment]
> Thanks. What other 32-bit apps that folks wanted to run on Nano Server container?

In our case it's old compilers we need to use, MinGW with gcc 4.8 and CodeSourcery with gcc for ARM 4.5. We have to guarantee to be able to rebuild our software with the same compiler and are not allowed to change them.

**Comment by mohsh86** [comment]
How come the issue is closed by providing a 64-bit exe version of a 32-bit one, when the issue clearly says add x86 support to nano image?

 A legacy piece of software few MB in size doesn't need the 5 GB+ servercore image!

**Comment by CeesvEgmond** [comment]
Here another request for 32bits enabled nano image.
We're using legacy simulation software written in Fortran. Compiling the source to 64 bits resulted in abnormal and unreliable results. 

For now we're using the 32 bits executable (10mb; command line) with a very large image.

**Comment by RealFrogo** [comment]
I'd like to add another reason to the one I mentioned almost a year ago, and which is still the most important issue for me. But recently I tried to use the nano server for something else, where we probably only need programs which all do exist in 64bit versions. However, what stopped me in my tracks this time was that some of them only come with an installer package that is 32bit based. That does seem contradictory at first, but is more common than I'd like.

So even when trying to only use 64bit versions, not being able to install them directly because of the 32 bit installers make things cumbersome again. And yes, of course there are workarounds, but I've got better things to do than trying to figure these out while maintaining our intended of-the-shelf approach with as little adjustment as possible, which is hindered by it. So my workaround for now is to stick to the oversized server core images after all.

**Comment by ntrappe-msft** [comment]
@deverouxmartin-msft  FYI.

**Comment by romandolinsky** [comment]
@akarshm why is these task close and complet ??
Are you providing requested functionalities? 
Can you provide links with solutions ?
In which major release we can expect solutions

**Comment by WillWilli** [comment]
+1 

Please reopen this issue as its not resolved.

**Comment by slonopotamus** [comment]
Ideally, there should be some `apt-get install` path that gradually increases functionality from a stripped-down `nanoserver` up to a full-blown `windows` image, so people can install stuff they need.

**Comment by neberson** [comment]
Do we have any news about this request to add 32bit support to nanoserver?

**Comment by akarshm** [comment]
We’re still working on it, but I don’t have a timeline to share at this moment.

**Comment by michbern-ms** [comment]
Good news!  As of the May Patch Tuesday, we now have Nano Server Features on Demand available, including 32-bit support.

https://techcommunity.microsoft.com/blog/containers/discover-the-new-era-of-windows-server-2025-nano-server-containers/4413060

**Comment by CeesvEgmond** [comment]
This is a great development @akarshm !
Although we cannot utilize this yet, we're running on Azure Container Instances and ltsc2025 is not supported there. ltsc2022 is supported but that does not have DISM.

Any solution here?

**Comment by slonopotamus** [comment]
Where we get a full list of available Features on Demand?

**Comment by michbern-ms** [comment]
> Where we get a full list of available Features on Demand?

The blog article linked above has a list of the main three features.   But you can also ask the system:

Run:
`docker run --user ContainerAdministrator -it mcr.microsoft.com/windows/nanoserver:ltsc2025 cmd.exe`

And from inside the container:
`dism /online /Get-Capabilities`
`dism /online /Get-CapabilityInfo /CapabilityName:Microsoft.NanoServer.OpenSSH.Server~~~~`

**Comment by neberson** [comment]
> This is a great development [@akarshm](https://github.com/akarshm) ! Although we cannot utilize this yet, we're running on Azure Container Instances and ltsc2025 is not supported there. ltsc2022 is supported but that does not have DISM.
> 
> Any solution here?

I have the same problem, when running ltsc2019 and ltsc2022 containers.

**Comment by akarshm** [comment]
Hi @neberson and @CeesvEgmond , as noted in the blog, this capability is only available in WS2025 (LTSC2025) containers.
